src/client-side-metrics/gcp-metrics-handler.ts

```diff
@@ -0,0 +1,189 @@
+            'The latencies of a client RPC attempt. Under normal circumstances, this value is identical to operation_latencies. If the client receives transient errors, however, then operation_latencies is the sum of all attempt_latencies and the exponential delays.',
+          unit: 'ms',
+        }),
+        retryCount: meter.createCounter('retry_count', {
+          description:
+            'A counter that records the number of attempts that an operation required to complete. Under normal circumstances, this value is empty.',
+        }),
+        applicationBlockingLatencies: meter.createHistogram(
+          'application_blocking_latencies',
+          {
+            description:
+              'The time from when the client receives the response to a request until the application reads the response. This metric is most relevant for ReadRows requests. The start and stop times for this metric depend on the way that you send the read request; see Application blocking latencies timer examples for details.',
+            unit: 'ms',
+          }
+        ),
+        firstResponseLatencies: meter.createHistogram(
+          'first_response_latencies',
+          {
+            description:
+              'Latencies from when a client sends a request and receives the first row of the response.',
+            unit: 'ms',
+          }
+        ),
+        serverLatencies: meter.createHistogram('server_latencies', {
+          description:
+            'Latencies between the time when the Google frontend receives an RPC and when it sends the first byte of the response.',
+        }),
+        connectivityErrorCount: meter.createHistogram(
+          'connectivity_error_count',
+          {
+            description:
+              "The number of requests that failed to reach Google's network. In normal cases, this number is 0. When the number is not 0, it can indicate connectivity issues between the application and the Google network.",
+          }
+        ),
+        clientBlockingLatencies: meter.createHistogram(
+          'client_blocking_latencies',
+          {
+            description:
+              'Latencies introduced when the client blocks the sending of more requests to the server because of too many pending requests in a bulk operation.',
+            unit: 'ms',
+          }
+        ),
+      };
+    }
+  }
+
+  /**
+   * Records metrics for a completed Bigtable operation.
+   * This method records the operation latency and retry count, associating them with provided attributes.
+   * @param {OnOperationCompleteMetrics} metrics Metrics related to the completed operation.
+   * @param {OnOperationCompleteAttributes} attributes Attributes associated with the completed operation.
+   */
+  onOperationComplete(
+    metrics: OnOperationCompleteMetrics,
+    attributes: OnOperationCompleteAttributes
+  ) {
+    this.initialize();
+    this.otelMetrics?.operationLatencies.record(
+      metrics.operationLatency,
+      attributes
```

Comment:
are you sure you attach all attributes?

What about potentially empty fields like zone?

Reply:
Yes. This is attached in the metrics collector. You can also see that it appears in the output in the `typical-method-call.txt` file.